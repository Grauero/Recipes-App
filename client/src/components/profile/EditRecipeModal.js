import React from 'react';
import PropTypes from 'prop-types';

const EditRecipeModal = ({ closeModal, handleChange }) => (
  <div className="modal modal-open">
    <div className="modal-inner">
      <div className="modal-content">
        <form className="modal-content-inner">
          <h4>Edit Recipe</h4>

          <label htmlFor="name">Recipe Name</label>
          <input type="text" name="name" onChange={handleChange} />

          <label htmlFor="imageUrl">Recipe Image</label>
          <input type="text" name="imageUrl" onChange={handleChange} />

          <label htmlFor="category">Category of Recipe</label>
          <select name="category" onChange={handleChange}>
            <option value="Breakfast">Breakfast</option>
            <option value="Lunch">Lunch</option>
            <option value="Dinner">Dinner</option>
            <option value="Snack">Snack</option>
          </select>

          <label htmlFor="description">Recipe Description</label>
          <input type="text" name="description" onChange={handleChange} />

          <hr />
          <div className="modal-buttons">
            <button type="submit" className="button-primary">
              Update
            </button>
            <button type="button" onClick={closeModal}>
              Cancel
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
);

EditRecipeModal.propTypes = {
  handleChange: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired
};

export default EditRecipeModal;
