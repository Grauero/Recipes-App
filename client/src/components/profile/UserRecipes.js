import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Query, Mutation } from 'react-apollo';
import { Link } from 'react-router-dom';

import EditRecipeModal from './EditRecipeModal';
import Spinner from '../common/Spinner';
import Error from '../common/Error';
import {
  GET_USER_RECIPES,
  DELETE_USER_RECIPE,
  GET_ALL_RECIPES,
  GET_CURRENT_USER
} from '../../queries';

class UserRecipes extends Component {
  state = {
    name: '',
    imageUrl: '',
    category: '',
    description: '',
    modal: false
  };

  handleChange = e => this.setState({ [e.target.name]: e.target.value });

  handleDelete = async (deleteUserRecipe) => {
    const deleteConfirmation = window.confirm('Are you sure you want to delete this recipe?');

    if (deleteConfirmation) {
      await deleteUserRecipe();
    }
  };

  updateCache = (cache, { data: { deleteUserRecipe } }) => {
    const { username } = this.props;
    const { getUserRecipes } = cache.readQuery({
      query: GET_USER_RECIPES,
      variables: { username }
    });

    cache.writeQuery({
      query: GET_USER_RECIPES,
      variables: { username },
      data: {
        getUserRecipes: getUserRecipes.filter(recipe => recipe._id !== deleteUserRecipe._id)
      }
    });
  };

  closeModal = () => this.setState({ modal: false });

  render() {
    const { username } = this.props;
    const { modal } = this.state;

    return (
      <Query query={GET_USER_RECIPES} variables={{ username }}>
        {(data, loading, error) => {
          if (loading) return <Spinner />;
          if (error) return <Error error={error} />;

          return (
            <ul>
              {modal && (
                <EditRecipeModal closeModal={this.closeModal} handleChange={this.handleChange} />
              )}
              {data.data && data.data.getUserRecipes && data.data.getUserRecipes.length !== 0 && (
                <h3>Your Recipes</h3>
              )}

              {data.data.getUserRecipes
                && data.data.getUserRecipes.map(recipe => (
                  <li key={recipe._id}>
                    <Link to={`/recipes/${recipe._id}`}>
                      <p>{recipe.name}</p>
                    </Link>
                    <p style={{ marginBottom: '0' }}>{recipe.likes}</p>
                    <Mutation
                      mutation={DELETE_USER_RECIPE}
                      variables={{ _id: recipe._id }}
                      refetchQueries={() => [
                        { query: GET_ALL_RECIPES },
                        { query: GET_CURRENT_USER }
                      ]}
                      update={this.updateCache}
                    >
                      {(deleteUserRecipe, attrs = {}) => (
                        <Fragment>
                          <button
                            type="button"
                            className="button-primary"
                            onClick={() => this.setState({ modal: true })}
                          >
                            Update
                          </button>
                          <p
                            className="delete-button"
                            onClick={() => this.handleDelete(deleteUserRecipe)}
                          >
                            {attrs.loading ? 'deleting...' : 'X'}
                          </p>
                        </Fragment>
                      )}
                    </Mutation>
                  </li>
                ))}
            </ul>
          );
        }}
      </Query>
    );
  }
}

UserRecipes.propTypes = {
  username: PropTypes.string.isRequired
};

export default UserRecipes;
