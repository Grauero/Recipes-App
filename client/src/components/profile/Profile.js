import React from 'react';
import PropTypes from 'prop-types';

import UserInfo from './UserInfo';
import UserRecipes from './UserRecipes';
import withAuth, { isAuth } from '../utils/withAuth';

const Profile = ({ session }) => (
  <div className="App">
    <UserInfo session={session} />
    <UserRecipes username={session.getCurrentUser.username} />
  </div>
);

Profile.propTypes = {
  session: PropTypes.shape({
    getCurrentUser: PropTypes.shape({
      username: PropTypes.string.isRequired
    }).isRequired
  }).isRequired
};

export default withAuth(isAuth)(Profile);
export { Profile };
