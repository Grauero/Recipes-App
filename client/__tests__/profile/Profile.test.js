import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { Profile } from '../../src/components/profile/Profile';

const props = { session: { getCurrentUser: { username: 'username', favorites: [] } } };
const component = shallow(<Profile {...props} />);

it('matches snapshot', () => {
  expect(toJSON(component)).toMatchSnapshot();
});
