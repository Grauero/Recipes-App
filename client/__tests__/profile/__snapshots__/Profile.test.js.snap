// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="App"
>
  <UserInfo
    session={
      Object {
        "getCurrentUser": Object {
          "favorites": Array [],
          "username": "username",
        },
      }
    }
  />
  <UserRecipes
    username="username"
  />
</div>
`;
